src/core/stdcpp/typeinfo.d

```diff
@@ -0,0 +1,146 @@
+// Written in the D programming language.
+
+/**
+ * Interface to C++ <typeinfo>
+ *
+ * Copyright: Copyright (c) 2016 D Language Foundation
+ * License:   $(WEB boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ * Authors:   $(WEB digitalmars.com, Walter Bright)
+ * Source:    $(DRUNTIMESRC core/stdcpp/_typeinfo.d)
+ */
+
+module core.stdcpp.typeinfo;
+
+version (CRuntime_DigitalMars)
+{
+    import core.stdcpp.exception;
+
+    extern (C++, std)
+    {
+        class type_info
+        {
+            void* pdata;
+
+          public:
+            //virtual ~this();
+            void dtor() { }     // reserve slot in vtbl[]
+
+            //bool operator==(const type_info rhs) const;
+            //bool operator!=(const type_info rhs) const;
+            final bool before(const type_info rhs) const;
+            final const(char)* name() const;
+          protected:
+            //type_info();
+          private:
+            //this(const type_info rhs);
+            //type_info operator=(const type_info rhs);
+        }
+
+        class bad_cast : core.stdcpp.exception.std.exception
+        {
+            this() nothrow { }
+            this(const bad_cast) nothrow { }
+            //bad_cast operator=(const bad_cast) nothrow { return this; }
+            //virtual ~this() nothrow;
+            override const(char)* what() const nothrow;
+        }
+
+        class bad_typeid : core.stdcpp.exception.std.exception
+        {
+            this() nothrow { }
+            this(const bad_typeid) nothrow { }
+            //bad_typeid operator=(const bad_typeid) nothrow { return this; }
+            //virtual ~this() nothrow;
+            override const (char)* what() const nothrow;
+        }
+    }
+}
+else version (CRuntime_Microsoft)
+{
+    import core.stdcpp.exception;
+
+    struct __type_info_node
+    {
+        void* _MemPtr;
+        __type_info_node* _Next;
+    }
+
+    extern __gshared __type_info_node __type_info_root_node;
+
+    extern (C++, std)
+    {
+        class type_info
+        {
+
+          public:
+            //virtual ~this();
+            void dtor() { }     // reserve slot in vtbl[]
+            //bool operator==(const type_info rhs) const;
+            //bool operator!=(const type_info rhs) const;
+            final bool before(const type_info rhs) const;
+            final const(char)* name(__type_info_node* p = &__type_info_root_node) const;
+
+          private:
+            void* pdata;
+            char _name[1];
+            //this(const type_info rhs);
+            //type_info operator=(const type_info rhs);
+        }
+
+        class bad_cast : core.stdcpp.exception.std.exception
+        {
+            this(const(char)* msg = "bad cast") { }
```

Comment:
As I said, D does not match C++ constructor mangling or ABI, so it shouldn't link, and if it did then it would corrupt memory and/or crash.  It could be made to work.